core/src/test/java/org/elasticsearch/common/util/concurrent/SuspendableRefContainerTests.java

```diff
@@ -0,0 +1,79 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util.concurrent;
+
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.nullValue;
+
+public class SuspendableRefContainerTests extends ESTestCase {
+
+    public void testBasicFunctionality() throws InterruptedException {
+        SuspendableRefContainer refContainer = new SuspendableRefContainer();
+        assertThat(refContainer.activeRefs(), equalTo(0));
+
+        Releasable lock1 = refContainer.tryAcquire();
+        assertThat(refContainer.activeRefs(), equalTo(1));
+        Releasable lock2 = refContainer.tryAcquire();
+        assertThat(refContainer.activeRefs(), equalTo(2));
+        lock1.close();
+        assertThat(refContainer.activeRefs(), equalTo(1));
+        lock2.close();
+        assertThat(refContainer.activeRefs(), equalTo(0));
+
+        Releasable block = refContainer.blockAcquisition();
+        assertThat(refContainer.activeRefs(), equalTo(0));
+        assertThat(refContainer.tryAcquire(), nullValue());
+        assertThat(refContainer.activeRefs(), equalTo(0));
+        block.close();
+        Releasable lock = refContainer.tryAcquire();
+        assertThat(refContainer.activeRefs(), equalTo(1));
+        lock.close();
+    }
+
+    public void testBlockingAcquisition() throws InterruptedException {
+        SuspendableRefContainer refContainer = new SuspendableRefContainer();
+        CountDownLatch latch = new CountDownLatch(1);
+
+        Thread t = new Thread(() -> {
+            try (Releasable releasable = refContainer.acquire()) {
+                latch.countDown();
+            } catch (InterruptedException e) {
+                fail("Interrupted");
+            }
+        });
+
+        try (Releasable block = refContainer.blockAcquisition()) {
+            t.start();
+            if (latch.await(100, TimeUnit.MILLISECONDS)) {
+                fail("reference should not have been acquired");
+            }
+        }
+
+        t.join(TimeValue.timeValueSeconds(30).millis());
```

Comment:
it's better (in general, here it's trivial) to not use timeouts as CI dumps a thread stack if a test doesn't finish within 20m. That tells us what the threads are doing. I use this timeouts too during development (and check that the thread stopped ;)). Can you remove the timeout?